### PR TITLE
Refactor ResteasyReactiveOutputStream so that it can be used by Quarkus CXF

### LINKED ISSUE #40994: Refactor `ResteasyReactiveOutputStream` so that it can be used by Quarkus CXF
### Description

`ResteasyReactiveOutputStream` was written so that RESTeasy is able to run on top of Vert.x. 
CXF works like RESTeasy: it also writes to a `(Servlet)OutputStream` and to make CXF able to run on top of Vert.x one needs a similar class. For some time, we had an adjusted static copy of  `ResteasyReactiveOutputStream` in Quarkus CXF. At some point, we got issues related to chunking reported, that were fixed in Quarkus long time ago. 
After that, I implemented a script that adapts `ResteasyReactiveOutputStream.java` at Quarkus CXF runtime which is a bit safer, but still rather fragile. 

### Implementation ideas

Hence, it would be nice if `ResteasyReactiveOutputStream` could be refactored in the following way:

* Make it independent of RESTeasy
* Possibly move it (along with its ancillary classes) to a separate module in Quarkus that depends only on `vertx-web` but does not depend on RESTeasy
* Or perhaps move the reafctored classes to Vert.x

@cescoffier encouraged me to prepare a PR which is underway.

### PR DESCRIPTION
**Pull request by ppalaga**
This PR is motivated by Quarkus CXF's need to have a class similar to `org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream`. Because it is not trivial to get right and currently cumbersome to keep in sync, this PR refactors `ResteasyReactiveOutputStream`, so that it can be re-used by Quarkus CXF and possibly other frameworks needing to pass HTTP responses via a `java.io.OutputStream`. 

In particular, this PR does the following:

* Make `ResteasyReactiveOutputStream` independent of RESTeasy
* Move it (along with its ancillary classes) to a separate module in Quarkus that depends only on `vertx-web` but does not depend on RESTeasy

If the new module will work well, we may consider moving it to Vert.x

- Resolves: #40994

### CODE REVIEW

**Comment by ppalaga** [comment]
Would you like to review @cescoffier ?

**Comment by ppalaga** [comment]
Not sure about the failure in `truststore-maven-plugin` on Windows. I see it passing on Linux. Is it perhaps intermittent?

Same with `MpRestClientAsyncTest.testIntegrationWithMpRestClient`: I cannot reproduce it locally. Is the test perhaps known to be flaky @gsmet @brunobat?

File: independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.util.CharsetUtil;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.impl.Arguments;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+public class VertxBufferImpl implements Buffer {
```
Comment (geoand) [inline independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:16]: Why is this needed? Don't we already have this class elsewhere?

**Comment by ppalaga** [comment]
> Please update the description with information about what this change enables.
> 
> Asking because for someone just seeing the PR out of the blue (or looking back at it a few months in the future), there is no context on why this was added.

No problem, done.

File: independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 1 above)
Comment (ppalaga) [reply independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:16]: Thanks for the hint. This one was supposed to be moved from `independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/VertxBufferImpl.java`, but I forgot to remove the old one. Which I am going to do right now.

There is still one more in `extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxBufferImpl.java` which can be removed in favor of this one. Which I am going to do as well.

Comment (geoand) [reply independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:16]: I would like to ask you to add a little Javadoc to the classes you are going to use in CXF, as these classes are now no longer going to be internal, so it would be nice for integrators to know a little about what each one does.

Comment (ppalaga) [reply independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:16]: Hm... now I see there is also `io.vertx.core.buffer.impl.BufferImpl` that does some additional bounds checks. I guess the two `VertxBufferImpl` do not do it intentionally for performance reasons?

Comment (ppalaga) [reply independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java]: > I would like to ask you to add a little Javadoc 

Haha, I can try

Comment (geoand) [reply independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java]: 🙏🏼

Comment (ppalaga) [reply independent-projects/resteasy-reactive/server/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java]: > I guess the two VertxBufferImpl do not do it intentionally for performance reasons?

OK, if so, I guess, it won't be so straightforward to move the refactored classes to Vert.x due to our "unsafe" `BufferImpl`. Therefore, I am going to rename the module to `quarkus-vertx-java-io` in line with the expectation that it will stay in Quarkus. Also, because the module is now used also in `quarkus-vertx` and `quarkus-websockets-next`, I am going to move it to directly under `independent-projects`.

**Comment by geoand** [comment]
cc @mkouba for the websockets-next change

File: independent-projects/vertx-java-io/pom.xml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../parent/pom.xml</relativePath>
+    </parent>
+
+    <groupId>io.quarkus.vertx-java-io</groupId>
+    <artifactId>quarkus-vertx-java-io</artifactId>
+    <name>Ancillary classes for using Vert.x with frameworks designed to read/write from/to java.io streams</name>
```
Comment (mkouba) [inline independent-projects/vertx-java-io/pom.xml:16]: I'm not so sure we should create an independent project for this kind of stuff. AFAIK there are some drawbacks related to version management, release process, etc.

@gsmet may know more.

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package io.quarkus.vertx.core.runtime;
+package io.quarkus.vertx.java.io;
```
Comment (mkouba) [inline independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:1]: I know that the content of `io.quarkus.vertx.core.runtime` is in theory not part of the public API but this looks like an unnecessary breaking change, or? Could we just extend a shared implementation instead? So that the clients do not need to change anything?

Maybe I'm complicating it too much...

File: independent-projects/vertx-java-io/pom.xml (hunk 2 above)
Comment (mkouba) [reply independent-projects/vertx-java-io/pom.xml:16]: Also `quarkus-vertx-java-io` feels really weird. I'd call it `quarkus-vertx-buffer`.

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java [hunk 4]
```diff
@@ -9,10 +9,14 @@
...
 import io.vertx.core.buffer.Buffer;
+import io.vertx.core.buffer.impl.BufferImpl;
 import io.vertx.core.impl.Arguments;
 import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 
+/**
+ * A variant of {@link BufferImpl} doing no bound checks for performance reasons.
+ */
 public class VertxBufferImpl implements Buffer {
```
Comment (mkouba) [inline independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:20]: Maybe we could also try to use a better name since it would be now part of the public API...

File: independent-projects/vertx-java-io/pom.xml (hunk 2 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/pom.xml:16]: > I'm not so sure we should create an independent project for this kind of stuff. AFAIK there are some drawbacks related to version management, release process, etc. @gsmet may know more.

+1 about the concern about the version management. I was hoping all was automated given that there are several independent projects already. Anyway, I am open to discuss where the module should be hosted. WDYT @gsmet?

> Also `quarkus-vertx-java-io` feels really weird. I'd call it `quarkus-vertx-buffer`.

Re-using the `VertxBufferImpl` was rather a secondary idea. The module was primarily intended to host the `VertxOutputStream` and perhaps also other java.io stream implementations usable for old frameworks wanting to run on top of Vert.x. I do not care much about the module name. `quarkus-vertx-java-io`, `quarkus-vertx-buffer` or even `quarkus-vertx-web-utils` would work for me.

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 3 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:1]: I'd let @geoand comment on that. From my outsider's PoV `VertxBufferImpl` seems to be a relatively simple and low-maintenance class, so we could also keep having two copies.

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 4 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:20]: Proposals?

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 3 above)
Comment (mkouba) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:1]: It's not about having two copies but about making the `io.quarkus.vertx.VertxBufferImpl` extend the new impl so that users of the original `VertxBufferImpl` are not broken...

File: independent-projects/vertx-java-io/pom.xml (hunk 2 above)
Comment (mkouba) [reply independent-projects/vertx-java-io/pom.xml:16]: > Re-using the VertxBufferImpl was rather a secondary idea. 

In that case something like `quarkus-vertx-utils` might make more sense...

BTW I also found `io.quarkus.resteasy.runtime.standalone.VertxOutputStream`. Isn't it a copy of the original `VertxOutputStream`?

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 4 above)
Comment (mkouba) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:20]: Yes, that's a little bit harder. `NoBoundChecksBufferImpl`? Maybe `QuarkusBufferImpl` - I saw this naming strategy here and there...

File: independent-projects/vertx-java-io/pom.xml (hunk 2 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/pom.xml:16]: > > Re-using the VertxBufferImpl was rather a secondary idea.
> 
> In that case something like `quarkus-vertx-utils` might make more sense...

OK, let me rename it to `quarkus-vertx-utils` unless nobody has a better idea.

> BTW I also found `io.quarkus.resteasy.runtime.standalone.VertxOutputStream`. Isn't it a copy of the original `VertxOutputStream`?

They are different. The VertxOutputStream from resteasy-reactive-vertx extends java.io.OutputStream rather than org.jboss.resteasy.spi.AsyncOutputStream

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 3 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:1]: I understand your concern with the public API and the proposed solution. I am just not 100% sure the gain of reusing the shared VertxBufferImpl by adding a dependency and letting the local VertxBufferImpl extend the shared one for the sake of backwards compatibility is worth the added complexity. I can do it like that if there are more voices in favor of that?

Comment (mkouba) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:1]: > ...is worth the added complexity

Well, the complexity here is just about keeping one more class. The dependency will be there anyway. But I don't insist it's just a proposal ;-).

Comment (geoand) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:1]: I don't think we need to be too concerned about this breakage.

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 4 above)
Comment (geoand) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:20]: `QuarkusBufferImpl` would be a good name, but this is a module in independent-projects :)

I would simply go with `NoBoundChecksBuffer`

File: independent-projects/vertx-java-io/pom.xml (hunk 2 above)
Comment (geoand) [reply independent-projects/vertx-java-io/pom.xml:16]: > I'm not so sure we should create an independent project for this kind of stuff. AFAIK there are some drawbacks related to version management, release process, etc.

I can't remember there being any such drawbacks

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 4 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:20]: Thanks for the feedback, let me rename it to `NoBoundChecksBuffer` then.

File: independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java (hunk 3 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/src/main/java/io/quarkus/vertx/java/io/VertxBufferImpl.java:1]: Thanks for the feedback, @mkouba and @geoand. Let me then leave this part as it is in the current state of the PR: +dependency -BufferImpl copy.

File: independent-projects/vertx-java-io/pom.xml (hunk 2 above)
Comment (ppalaga) [reply independent-projects/vertx-java-io/pom.xml:16]: > > I'm not so sure we should create an independent project for this kind of stuff. AFAIK there are some drawbacks related to version management, release process, etc.
> 
> I can't remember there being any such drawbacks

There are some version properties in various independent projects, which have to be kept in sync on various places. This PR makes the mess a bit worse by adding two new properties in the new module.
It would be nice to hear what @gsmet thinks.

**Review by geoand — APPROVED** [review]

**Comment by brunobat** [comment]
> Not sure about the failure in `truststore-maven-plugin` on Windows. I see it passing on Linux. Is it perhaps intermittent?
> 
> Same with `MpRestClientAsyncTest.testIntegrationWithMpRestClient`: I cannot reproduce it locally. Is the test perhaps known to be flaky @gsmet @brunobat?

That's been failing a bit. Likely related.

**Review by gsmet — APPROVED** [review]
We can merge this but I would like us to start the discussion to move this to Vert.x as I would like us to get rid of it at some point in the future.

**Comment by ppalaga** [comment]
> We can merge this but I would like us to start the discussion to move this to Vert.x as I would like us to get rid of it at some point in the future.

Thanks for merging, @gsmet, I have created a followup We can merge this but I would like us to start the discussion to move this to Vert.x as I would like us to get rid of it at some point in the future.
